python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -652,3 +658,220 @@ def _isin_datetimelike(
+        Left bound for generating dates.
+
+    end : str or datetime-like, optional
+        Right bound for generating dates.
+
+    periods : int, optional
+        Number of periods to generate.
+
+    freq : DateOffset
+        Frequencis to generate the datetime series. Mixed fixed-frequency and
+        non-fixed frequency offset is unsupported. See notes for detail.
+
+    tz : str or tzinfo, optional
+        Not Supported
+
+    normalize : bool, default False
+        Not Supported
+
+    name : str, default None
+        Name of the resulting DatetimeIndex
+
+    closed : {None, 'left', 'right'}, optional
+        Not Supported
+
+    Returns
+    -------
+    DatetimeIndex
+
+    Notes
+    -----
+    Of the four parameters `start`, `end`, `periods`, and `freq`, exactly three
+    must be specified. If `freq` is omitted, the resulting DatetimeIndex will
+    have periods linearly spaced elements between start and end (closed on both
+    sides).
+
+    cudf supports `freq` specified with either fixed-frequency offset
+    (such as weeks, days, hours, minutes...) or non-fixed frequency offset
+    (such as years and months). Specifying `freq` with a mixed fixed and
+    non-fixed frequency is currently unsupported. For example:
+
+    >>> cudf.date_range(
+    ...     start='2021-08-23 08:00:00',
+    ...     freq=cudf.DateOffset(months=2, days=5),
+    ...     periods=5)
+    ...
+    NotImplementedError: Mixing fixed and non-fixed frequency offset is
+    unsupported.
+
+    Examples
+    --------
+    >>> cudf.date_range(
+    ...     start='2021-08-23 08:00:00',
+    ...     freq=cudf.DateOffset(years=1, months=2),
+    ...     periods=5)
+    DatetimeIndex(['2021-08-23 08:00:00', '2022-10-23 08:00:00',
+                '2023-12-23 08:00:00', '2025-02-23 08:00:00',
+                '2026-04-23 08:00:00'],
+                dtype='datetime64[ns]')
+
+    """
+    if tz is not None:
+        raise NotImplementedError("tz is currently unsupported.")
+
+    if normalize:
+        raise NotImplementedError(
+            "normalizing timestamps to midnight is unsupported."
+        )
+
+    if closed is not None:
+        raise NotImplementedError("closed is currently unsupported.")
+
+    if [start, end, periods, freq].count(None) > 1:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and freq, exactly "
+            "three must be specified"
+        )
+
+    dtype = np.dtype("<M8[ns]")
+
+    _periods_not_specified = False
+
+    if freq is not None:
+        if isinstance(freq, DateOffset):
+            offset = freq
+        else:
+            raise TypeError("`freq` must be a cudf.DateOffset object.")
+
+        if _check_mixed_freqeuency(offset):
+            raise NotImplementedError(
+                "Mixing fixed and non-fixed frequency offset is unsupported."
+            )
+        offset = offset.to_pandas()
+
+        if start is None:
+            end = cudf.Scalar(end, dtype=dtype)
```

Comment:
if we work in the `datetime64[ns]` range exclusively, do we need to restrict the user from passing datetimes outside the bounds? I am a little concerned about overflow here. Pandas checks but we error for a different reason:

```
>>> start = np.datetime64(np.iinfo('int64').max, 'ns').astype('datetime64[ms]')
>>> end = start + np.timedelta64(100, 'ms')
>>> start
numpy.datetime64('2262-04-11T23:47:16.854')
>>> end
numpy.datetime64('2262-04-11T23:47:16.954')
>>> pd.date_range(start, end, freq='ms')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nfs/brmiller/anaconda3/envs/pr_8886/lib/python3.8/site-packages/pandas/core/indexes/datetimes.py", line 1069, in date_range
    dtarr = DatetimeArray._generate_range(
  File "/home/nfs/brmiller/anaconda3/envs/pr_8886/lib/python3.8/site-packages/pandas/core/arrays/datetimes.py", line 383, in _generate_range
    end = Timestamp(end)
  File "pandas/_libs/tslibs/timestamps.pyx", line 1095, in pandas._libs.tslibs.timestamps.Timestamp.__new__
  File "pandas/_libs/tslibs/conversion.pyx", line 403, in pandas._libs.tslibs.conversion.convert_to_tsobject
  File "pandas/_libs/tslibs/conversion.pyx", line 190, in pandas._libs.tslibs.conversion.get_datetime64_nanos
  File "pandas/_libs/tslibs/np_datetime.pyx", line 113, in pandas._libs.tslibs.np_datetime.check_dts_bounds
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 2262-04-11 23:47:16

>>> cudf.date_range(start, end, freq='ms')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nfs/brmiller/anaconda3/envs/pr_8886/lib/python3.8/site-packages/cudf/core/tools/datetimes.py", line 769, in date_range
    raise TypeError("`freq` must be a cudf.DateOffset object.")
TypeError: `freq` must be a cudf.DateOffset object.
>>> cudf.date_range(start, end, freq=cudf.DateOffset(milliseconds=1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nfs/brmiller/anaconda3/envs/pr_8886/lib/python3.8/site-packages/cudf/core/tools/datetimes.py", line 788, in date_range
    int(end - start) / _offset_to_nanoseconds_lower_bound(offset)
ZeroDivisionError: float division by zero
```

Reply:
This div by zero issue happens to result not from nanoseconds overflow, but because pd.DateOffset constructed with milliseconds is discarded. And so the offset is left as zero. Should we disable `milliseconds` support in cudf to match pandas behavior?

As for nanoseconds overflow, in fixed frequency case, we always have to go through this line:
```python
end = (pd.Timestamp(start.value) + (periods - 1) * offset).to_numpy()
```
to construct a pandas timestamp type for the last element in the sequence. Since both end of the sequence has to go through pd.Timestamp, if there's any overflow in the sequence it's guaranteed to be thrown at this step.

However, for non-fixed frequency case, since computation goes through libcudf, we might encounter a silence overflow:

```python
import cudf, numpy as np, pandas as pd
start = np.datetime64(np.iinfo('int64').max, 'ns').astype('datetime64[us]')
cudf.date_range(start, periods=2, freq=cudf.DateOffset(months=1))

DatetimeIndex(['2262-04-11 23:47:16.854775', '1677-10-21 00:12:43.145223384'], dtype='datetime64[ns]')
```

Reply:
Result from offline discussion - will move the `end` computation to before calculating any sequences. This makes sure overflows are thrown.

Reply:
I am +1 on that solution.